File: c/1079.c
void sort(char* a, int n) {
    for (int i = 0; i < n - 1; i++) {
        int t = i;
        for (int j = i + 1; j < n; j++)
            if (a[t] > a[j])
                t = j;
        char s = a[i];
        a[i] = a[t];
        a[t] = s;
    }
}

int count(char* c, bool* trav, int n) {
    int cc = 0;
    for (int i = 0; i < n; i++)
        if (!trav[i]) {
            trav[i] = true;
            cc += count(c, trav, n) + 1;
            trav[i] = false;
            int t = i;
            while (i < n && c[t] == c[i])
                ++i;
            --i;
        }
    return cc;
}

int numTilePossibilities(char* tiles) {
    int n = strlen(tiles);
    sort(tiles, n);
    bool trav[8] = {false};
    return count(tiles, trav, n);
}

File: javascript/1079.js
/**
 * @param {string} tiles
 * @return {number}
 */
var numTilePossibilities = function (tiles) {
    var count = function (c, trav, n) {
        let cc = 0;
        for (let i = 0; i < n; i++)
            if (!trav[i]) {
                trav[i] = true;
                cc += count(c, trav, n) + 1;
                trav[i] = false;
                let t = i;
                while (i < n && c[t] == c[i])
                    ++i;
                --i;
            }
        return cc;
    };
    let c = []
    for (let i of tiles)
        c.push(i)
    c.sort()
    let n = c.length;
    let trav = new Array(n).fill(false)
    return count(c, trav, n);
};
